### PR TITLE
[docs] callbacks and collators

### PR DESCRIPTION
**Pull request by stevhliu**
part 2 of refactoring the training docs adds new dedicated guide to callbacks and data collators

todo:

- [x] backlink to `## Next steps` in `trainer.md` once https://github.com/huggingface/transformers/pull/44185 is merged

### CODE REVIEW

File: docs/source/en/data_collators.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+Pass the custom data collator to [`Trainer`] like any other data collator.
+
+```py
+trainer = Trainer(
+    ...,
+    data_collator=DataCollatorWithScore(tokenizer=tokenizer),
+)
+```
+
+## DataCollatorMixin
```
Comment (stevhliu) [inline docs/source/en/data_collators.md:74]: @qgallouedec if you could review this section, that'd be great thanks! 🙂

File: docs/source/en/data_collators.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+-->
+
+# Data collators
+
+A data collator assembles individual dataset samples into a batch for the model. It can also dynamically pad samples to the longest sequence in *each batch*, which is more efficient than padding to a global maximum length.
+
+```md
+Dataset[0] → {"input_ids": [101, 2003, ...], "labels": 1}
+Dataset[1] → {"input_ids": [101, 2003, 1996, ...], "labels": 0}
+Dataset[2] → {"input_ids": [101, 7592, ...], "labels": 1}
```
Comment (qgallouedec) [inline docs/source/en/data_collators.md]:
```suggestion
Dataset[0] → {"input_ids": [101, 2003], "labels": 1}
Dataset[1] → {"input_ids": [101, 2003, 1996], "labels": 0}
Dataset[2] → {"input_ids": [101, 7592], "labels": 1}
```

File: docs/source/en/data_collators.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+from transformers import DataCollatorMixin
+from trl.trainer.utils import pad
+
+class DataCollatorForPreference(DataCollatorMixin):
+    pad_token_id: int
+    pad_to_multiple_of: int | None = None
+
+    def __call__(self, examples: list[dict]) -> dict:
+        chosen_input_ids   = [torch.tensor(ex["chosen_input_ids"])   for ex in examples]
+        rejected_input_ids = [torch.tensor(ex["rejected_input_ids"]) for ex in examples]
```
Comment (qgallouedec) [inline docs/source/en/data_collators.md]:
```suggestion
        chosen_input_ids   = [torch.tensor(ex["chosen_ids"])   for ex in examples]
        rejected_input_ids = [torch.tensor(ex["rejected_ids"]) for ex in examples]
```

nit, we recently renamed these

File: docs/source/en/data_collators.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+```py
+trainer = Trainer(
+    ...,
+    data_collator=DataCollatorWithScore(tokenizer=tokenizer),
+)
+```
+
+## DataCollatorMixin
+
+Subclass [`DataCollatorMixin`] for full control over batch assembly and implement your own `__call__` method. Build custom padding logic, handle multiple input types, or create entirely new batch structures. The [`~trl.trainer.dpo_trainer.DataCollatorForPreference`] example below uses [`DataCollatorMixin`] because each training sample has a chosen and rejected response, and the model needs to see both.
```
Comment (qgallouedec) [inline docs/source/en/data_collators.md]: 1. we should soon remove `DataCollatorForPreference` from the pubic API, meaning it will continue to exist in the repo, but won't appear in the documentation, see https://github.com/huggingface/trl/issues/5090. I recommend linking the GitHub lines instead. 
2. To be closer to the example below, maybe you can point to this collator instead https://github.com/huggingface/trl/blob/fa313fd57244008953753047795c954a782f9cfc/trl/trainer/reward_trainer.py#L126

**Review by qgallouedec — APPROVED** [review]
nice!
I've read the whole data collator section and made a few comments
I've read quickly the callback section, and it looks good to me :)

File: docs/source/en/trainer_callbacks.md [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+          on_optimizer_step       ← after optimizer.step()
+          on_step_end
+          └─ (conditionally):
+              on_log
+              on_evaluate
+              on_save
+      on_epoch_end
+on_train_end
+on_predict / on_prediction_step
+on_push_begin
```
Comment (SunMarc) [inline docs/source/en/trainer_callbacks.md:43]: nice way to showcase all the event !

File: docs/source/en/trainer_callbacks.md [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+```python
+trainer = Trainer(
+    callbacks=[EpochLoggerCallback(lr_eval_threshold=1e-5)],
+    ...,
+)
+```
+
+## Built-in callbacks
+
+Transformers includes several built-in callbacks that are active by default. Additional [integrated callbacks](./main_classes/callback#available-callbacks) log to platforms like [Trackio](https://huggingface.co/docs/trackio/en/index) and [CodeCarbon](https://mlco2.github.io/codecarbon/).
```
Comment (SunMarc) [inline docs/source/en/trainer_callbacks.md]: it's fine to remove codecarbon mention. Can you also update the api page as things changed a bit ? ./main_classes/callback#available-callbacks 
Since v5, the default for `report_to` is `none`.

File: docs/source/en/trainer_callbacks.md [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+### ProgressCallback
+
+[`ProgressCallback`] displays a progress bar during training and a separate bar during evaluation or prediction. On each `on_log` event, it prints the latest metrics alongside the bar. During distributed training, it only runs on the main process to avoid duplicate output.
+
+Replace with [`PrinterCallback`] if you want printed logs instead of a progress bar.
+
+### PrinterCallback
+
+[`PrinterCallback`] prints the log dictionary to stdout on every `on_log` event with no progress bar.
+
```
Comment (SunMarc) [inline docs/source/en/trainer_callbacks.md]: maybe we can specify the following: 

(the first one is used if you deactivate tqdm through the [TrainingArguments](https://huggingface.co/docs/transformers/v5.2.0/en/main_classes/trainer#transformers.TrainingArguments), otherwise it’s the second one).

File: docs/source/en/trainer_callbacks.md [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+Use [`~Trainer.remove_callback`] and [`~Trainer.add_callback`] to replace a callback.
+
+```py
+from transformers import PrinterCallback
+
+trainer = Trainer(...)
+trainer.remove_callback(ProgressCallback)
+trainer.add_callback(PrinterCallback)
+```
```
Comment (SunMarc) [inline docs/source/en/trainer_callbacks.md]: we can maybe move that above ?

File: docs/source/en/trainer_callbacks.md [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+### ProgressCallback
+
+[`ProgressCallback`] displays a progress bar during training and a separate bar during evaluation or prediction. On each `on_log` event, it prints the latest metrics alongside the bar. During distributed training, it only runs on the main process to avoid duplicate output.
+
+Replace with [`PrinterCallback`] if you want printed logs instead of a progress bar.
+
+### PrinterCallback
+
+[`PrinterCallback`] prints the log dictionary to stdout on every `on_log` event with no progress bar.
```
Comment (SunMarc) [inline docs/source/en/trainer_callbacks.md]: we can merge those two together if this makes more sense

**Review by SunMarc — APPROVED** [review]
Thanks a lot ! added a couple of feedback

File: docs/source/en/trainer_callbacks.md (hunk 9 above)
Comment (stevhliu) [reply docs/source/en/trainer_callbacks.md]: this is the way!
